File: examples/openobserve/03_hardening_profile.py
#!/usr/bin/env python3
# SPDX-FileCopyrightText: Copyright (C) 2026 provide.io llc
# SPDX-License-Identifier: Apache-2.0
# SPDX-Comment: Part of provide-telemetry.
#

from __future__ import annotations

import os
import time
from base64 import b64encode
from urllib.parse import quote

from provide.telemetry import (
    PIIRule,
    event_name,
    get_health_snapshot,
    get_logger,
    register_cardinality_limit,
    register_pii_rule,
    setup_telemetry,
    shutdown_telemetry,
    trace,
)
from provide.telemetry.config import TelemetryConfig


def _require_env(name: str) -> str:
    value = os.getenv(name)
    if not value:
        msg = f"missing required env var: {name}"
        raise RuntimeError(msg)
    return value


@trace(event_name("example", "openobserve", "work"))
def _emit(iteration: int) -> None:
    token_value = os.getenv("PROVIDE_EXAMPLE_TOKEN", "example-token-from-env")
    get_logger("examples.openobserve.hardening").info(
        event_name("example", "openobserve", "log"),
        iteration=iteration,
        user={"email": "ops@example.com", "full_name": "Operator Example"},
        token=token_value,
    )


def main() -> None:
    base_url = _require_env("OPENOBSERVE_URL").rstrip("/")
    user = _require_env("OPENOBSERVE_USER")
    password = _require_env("OPENOBSERVE_PASSWORD")

    auth = f"Basic {b64encode(f'{user}:{password}'.encode()).decode('ascii')}"
    cfg = TelemetryConfig.from_env(
        {
            "PROVIDE_TELEMETRY_SERVICE_NAME": "provide-telemetry-hardening-example",
            "PROVIDE_TELEMETRY_VERSION": "hardening",
            "PROVIDE_SAMPLING_LOGS_RATE": "1.0",
            "PROVIDE_SAMPLING_TRACES_RATE": "1.0",
            "PROVIDE_SAMPLING_METRICS_RATE": "1.0",
            "PROVIDE_BACKPRESSURE_TRACES_MAXSIZE": "64",
            "PROVIDE_EXPORTER_LOGS_RETRIES": "1",
            "PROVIDE_EXPORTER_TRACES_RETRIES": "1",
            "PROVIDE_EXPORTER_METRICS_RETRIES": "1",
            "OTEL_EXPORTER_OTLP_HEADERS": f"Authorization={quote(auth, safe='')}",
            "OTEL_EXPORTER_OTLP_TRACES_ENDPOINT": f"{base_url}/v1/traces",
            "OTEL_EXPORTER_OTLP_METRICS_ENDPOINT": f"{base_url}/v1/metrics",
            "OTEL_EXPORTER_OTLP_LOGS_ENDPOINT": f"{base_url}/v1/logs",
            "PROVIDE_SLO_ENABLE_RED_METRICS": "true",
            "PROVIDE_SLO_ENABLE_USE_METRICS": "true",
        }
    )

    register_pii_rule(PIIRule(path=("user", "email"), mode="hash"))
    register_pii_rule(PIIRule(path=("user", "full_name"), mode="truncate", truncate_to=4))
    register_cardinality_limit("player_id", max_values=50, ttl_seconds=300)

    setup_telemetry(cfg)

    for i in range(5):
        _emit(i)
        time.sleep(0.05)

    shutdown_telemetry()
    print({"health": get_health_snapshot()})


if __name__ == "__main__":
    main()

File: typescript/examples/openobserve/03_hardening_profile.ts
// SPDX-FileCopyrightText: Copyright (C) 2026 provide.io llc
// SPDX-License-Identifier: Apache-2.0
// SPDX-Comment: Part of Provide Telemetry.

/**
 * OpenObserve hardening profile — PII masking, cardinality, sampling,
 * backpressure, exporter resilience, and RED/USE SLO metrics, all active.
 *
 * Required env vars:
 *   OPENOBSERVE_URL      e.g. http://localhost:5080/api/default
 *   OPENOBSERVE_USER     e.g. admin@provide.test
 *   OPENOBSERVE_PASSWORD e.g. Complexpass#123
 *
 * Run:
 *   OPENOBSERVE_URL=http://localhost:5080/api/default \
 *   OPENOBSERVE_USER=admin@provide.test \
 *   OPENOBSERVE_PASSWORD=Complexpass#123 \
 *   npx tsx examples/openobserve/03_hardening_profile.ts
 */

import {
  eventName,
  getConfig,
  getHealthSnapshot,
  getLogger,
  recordRedMetrics,
  recordUseMetrics,
  registerCardinalityLimit,
  registerOtelProviders,
  registerPiiRule,
  setExporterPolicy,
  setQueuePolicy,
  setSamplingPolicy,
  setupTelemetry,
  shutdownTelemetry,
  withTrace,
} from '../../src/index.js';

function requireEnv(name: string): string {
  const val = process.env[name];
  if (!val) throw new Error(`missing required env var: ${name}`);
  return val;
}

function authHeader(user: string, password: string): string {
  return `Basic ${Buffer.from(`${user}:${password}`).toString('base64')}`;
}

async function main(): Promise<void> {
  const baseUrl = requireEnv('OPENOBSERVE_URL').replace(/\/$/, '');
  const user = requireEnv('OPENOBSERVE_USER');
  const password = requireEnv('OPENOBSERVE_PASSWORD');
  const auth = authHeader(user, password);

  // ── Hardening guardrails (before setup, so they're active from the start) ──

  registerPiiRule({ path: 'user.email', mode: 'hash' });
  registerPiiRule({ path: 'user.full_name', mode: 'truncate', truncateTo: 4 });
  registerCardinalityLimit('player_id', { maxValues: 50, ttlSeconds: 300 });

  // Per-signal sampling — matches Python's PROVIDE_SAMPLING_*_RATE=1.0
  setSamplingPolicy('logs', { defaultRate: 1.0 });
  setSamplingPolicy('traces', { defaultRate: 1.0 });
  setSamplingPolicy('metrics', { defaultRate: 1.0 });

  setQueuePolicy({ maxLogs: 0, maxMetrics: 0, maxTraces: 64 });

  // Per-signal resilience — matches Python's PROVIDE_EXPORTER_*_RETRIES=1
  setExporterPolicy('logs', { retries: 1, backoffMs: 0, failOpen: true, timeoutMs: 5000 });
  setExporterPolicy('traces', { retries: 1, backoffMs: 0, failOpen: true, timeoutMs: 5000 });
  setExporterPolicy('metrics', { retries: 1, backoffMs: 0, failOpen: true, timeoutMs: 5000 });

  // ── @provide-io/telemetry setup ───────────────────────────────────────────

  setupTelemetry({
    serviceName: 'provide-telemetry-hardening-example',
    logLevel: 'debug',
    consoleOutput: false,
    otelEnabled: true,
    otlpEndpoint: baseUrl,
    otlpHeaders: { Authorization: auth },
    environment: 'development',
    version: 'hardening',
    strictSchema: true,
  });
  await registerOtelProviders(getConfig());

  const log = getLogger('examples.openobserve.hardening');
  const tokenValue = process.env['PROVIDE_EXAMPLE_TOKEN'] ?? 'example-token-from-env';

  // ── Emit with hardening active ───────────────────────────────────────────

  for (let i = 0; i < 5; i++) {
    await withTrace(event('example', 'openobserve', 'work').event, async () => {
      log.info({
        ...event('example', 'openobserve', 'log'),
        iteration: i,
        user: { email: 'ops@example.com', full_name: 'Operator Example' },
        token: tokenValue,
      });
      recordRedMetrics({ route: '/example', method: 'GET', statusCode: 200, durationMs: 10 + i });
      recordUseMetrics({ resource: 'cpu', utilization: 40 + i * 5 });
    });
    await new Promise((r) => setTimeout(r, 50));
  }

  // ── Flush and shutdown ───────────────────────────────────────────────────

  await shutdownTelemetry();

  const health = getHealthSnapshot();
  console.log(JSON.stringify({ health }));
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});
